README.md
<p align="center">
  <img width="50%" height="auto" src="https://vignette.wikia.nocookie.net/logopedia/images/2/23/Admob12.png/revision/latest?cb=20190109132636">
</p>

# [Capacitor AdMob](https://github.com/rahadur/capacitor-admob) 💸

Capacitor AdMob is a native AdMob implementation for IOS & Android. Now you can use this package as a [Ionic Capacitor](https://capacitor.ionicframework.com) Plugin in your App.

## Changelog:

### [v1.0.0](https://github.com/rahadur/capacitor-admob/tree/v1.0.0)
- Fixed Androidx CoordinatorLayout **[#35](https://github.com/rahadur/capacitor-admob/issues/35)**
- Fixed Banner and interstitial share the same Event Listeners **[#40](https://github.com/rahadur/capacitor-admob/issues/40)**
- Fixed AdMob error in Mobile after build **[#48](https://github.com/rahadur/capacitor-admob/issues/48)**
- Fixed Ad overlaps tab bar iOS **[#27](https://github.com/rahadur/capacitor-admob/issues/27)**
- Migrate Demo App to Capacitor 2.1.2

#### [Previous Release](./CHANGELOG.md)


## Supported Platform

- [x] iOS
- [x] Android

## Other Plugins:

| Plugins                                             | Android | iOS | Electron | PWA |
| :-------------------------------------------------- | :------ | :-- | :------- | :-- |
| [MoPub](https://github.com/rahadur/capacitor-mopub) | ✅      | ❌  | ❌       | ❌  |

# 💰 For a Few Dollars More

Thanks for considering donate.

If this plugin help you to earn few dollars, please don't forget to share your profit, because **"Sharing is Caring"**. Your donation will inspire me a lot to give more time on open-source project & help you by creating more useful plugins.

| Methond   | Type    | Amount | Link                                                                                             |
| :-------- | :------ | :----- | :----------------------------------------------------------------------------------------------- |
| Paypal.me | Once    | Any    | [paypal.me](https://www.paypal.me/xplatform)                                                     |
| Paypal    | Monthly | \$5    | [paypal.com](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=76FEY9XUD9A98) |
| Paypal    | Monthly | \$10   | [paypal.com](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=HUQLAF8W5M2FA) |
| Paypal    | Monthly | \$25   | [paypal.com](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=RWELCRVN64GHC) |
| Paypal    | Monthly | \$50   | [paypal.com](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=7C4EEH6ELENKA) |

## AdMob Demo App

Download Demo App from **[Here](https://github.com/rahadur/capacitor-admob/tree/master/admob-demo)**

## Screenshot

### iOS

| Basic Banner AD                                                  | Interstitial AD                                                                    | Interstitial Video AD                                                                               | Reward Video AD                                                               |
| ---------------------------------------------------------------- | ---------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------- |
| ![Alt text](/screenshort/ios/banner_ad.png?raw=true "Banner AD") | ![Intertitial AD ](/screenshort/ios/intertitial_ad.png?raw=true "Intertitial AD ") | ![Intertitial Video AD](/screenshort/ios/intertitial_video_ad.png?raw=true "Intertitial Video AD ") | ![Reward Video AD](/screenshort/ios/reward_ad.png?raw=true "Reward Video AD") |

### Android

| Basic Banner AD                                                      | Interstitial AD                                                                        | Interstitial Video AD                                                                                   | Reward Video AD                                                                   |
| -------------------------------------------------------------------- | -------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------- |
| ![Alt text](/screenshort/android/banner_ad.png?raw=true "Banner AD") | ![Intertitial AD ](/screenshort/android/intertitial_ad.png?raw=true "Intertitial AD ") | ![Intertitial Video AD](/screenshort/android/intertitial_video_ad.png?raw=true "Intertitial Video AD ") | ![Reward Video AD](/screenshort/android/reward_ad.png?raw=true "Reward Video AD") |

```console
cd admob-demo

npm install

ionic build

npx cap copy

npx cap sync

npx cap update

npx cap open android

============== Or just use this command ===========

npm install & ionic build & npx cap copy & npx cap sync & npx cap update & npx cap open android
```

## Installation

Use **AdMob** plugins in your app.

```console
 npm install --save capacitor-admob
```

## iOS

### Update **Info.plist**

Open your **App/App/Info.plist** file and add this `plist value` line at the right spot (and replace the value by the actual App ID of your app!):

```xml
<!-- this two line needs to be added -->

<key>GADIsAdManagerApp</key>
<true/>

<key>GADApplicationIdentifier</key>
<!-- replace this value with your App ID key-->
<string>ca-app-pub-6564742920318187~7217030993</string>

```

## Android

### Update Manifest

Open your **android/app/src/Android/AndroidManifest.xml** file and add this `meta-data` line at the right spot (and replace the value by the actual App ID of your app!):

```xml
<application>
  <!-- this line needs to be added (replace the value!) -->
  <meta-data android:name="com.google.android.gms.ads.APPLICATION_ID" android:value="ca-app-pub-3940256099942544~3347511713" />
  <activity></activity>
</application>

```

### Register AdMob to Capacitor Android

Open your Ionic Capacitor App in Android Studio, Now open **MainActivity.java** of your app and Register AdMob to Capacitor Plugins.

```java
// Other imports...
import app.xplatform.capacitor.plugins.AdMob;

public class MainActivity extends BridgeActivity {
  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    this.init(savedInstanceState, new ArrayList<Class<? extends Plugin>>() {{

      add(AdMob.class);  // Add AdMob as a Capacitor Plugin

    }});
  }
}
```

## 📌 Initialize AdMob

Open your Ionic app **app.component.ts** file and add this folloing code.

```typescript
import { Plugins } from "@capacitor/core";
// import { initialize } from 'capacitor-admob'; No longar required

const { AdMob } = Plugins;

@Component({
  selector: "app-root",
  templateUrl: "app.component.html",
  styleUrls: ["app.component.scss"]
})
export class AppComponent {
  constructor() {
    // Initialize AdMob for your Application
    AdMob.initialize("YOUR APPID");
  }
}
```

## 📌 BANNER

### showBanner(options: AdOptions): Promise<{ value: boolean }>

```typescript
import { Plugins } from "@capacitor/core";
import { AdOptions, AdSize, AdPosition } from "capacitor-admob";

const { AdMob } = Plugins;

@Component({
  selector: "admob",
  templateUrl: "admob.component.html",
  styleUrls: ["admob.component.scss"]
})
export class AdMobComponent {
  options: AdOptions = {
    adId: "YOUR ADID",
    adSize: AdSize.BANNER,
    position: AdPosition.BOTTOM_CENTER
  };

  constructor() {
    // Show Banner Ad
    AdMob.showBanner(this.options).then(
      value => {
        console.log(value); // true
      },
      error => {
        console.error(error); // show error
      }
    );

    // Subscibe Banner Event Listener
    AdMob.addListener("onAdLoaded", (info: boolean) => {
      console.log("Banner Ad Loaded");
    });
  }
}
```

### **Show TabBar Banner AD**

```typescript
import { Plugins } from "@capacitor/core";
import { AdOptions, AdSize, AdPosition } from "capacitor-admob";

const { AdMob } = Plugins;

@Component({
  selector: "admob",
  templateUrl: "admob.component.html",
  styleUrls: ["admob.component.scss"]
})
export class AdMobComponent {
  options: AdOptions = {
    adId: 'ca-app-pub-3940256099942544/6300978111',
    adSize: AdSize.SMART_BANNER,
    position: AdPosition.BOTTOM_CENTER,
    hasTabBar: true,  // make it true if you have TabBar Layout.
    tabBarHeight: 56  // you can assign custom margin in pixel default is 56
  };

  constructor() {
    // Show TabBar Banner Ad
    AdMob.showBanner(options)
        .then(
            async (value) => {
              console.log(value);  // true
              await Toast.show({
                text: 'Showing TabBar Banner AD.'
              })
            },
            (error) => {
              console.error(error); // show error
            }
        );


    // Subscribe Banner Event Listener
    AdMob.addListener('onAdLoaded', async (info: boolean) => {
      console.log('Showing TabBar Banner AD.');
    });
  }
}
```

### hideBanner(): Promise<{ value: boolean }>

```typescript
// Hide the banner, remove it from screen, but can show it later

AdMob.hideBanner().then(
  value => {
    console.log(value); // true
  },
  error => {
    console.err(error); // show error
  }
);
```

### resumeBanner(): Promise<{ value: boolean }>

```typescript
// Resume the banner, show it after hide

AdMob.resumeBanner().then(
  value => {
    console.log(value); // true
  },
  error => {
    console.error(error); // show error
  }
);
```

### removeBanner(): Promise<{ value: boolean }>

```typescript
// Destroy the banner, remove it from screen.

AdMob.removeBanner().then(
  value => {
    console.log(value); // true
  },
  error => {
    console.error(error); // show error
  }
);
```

### Event Listener

This following Event Listener can be called in **Banner AD**.

```typescript
addListener(eventName: 'onAdLoaded', listenerFunc: (info: any) => void): PluginListenerHandle;

addListener(eventName: 'onAdFailedToLoad', listenerFunc: (info: any) => void): PluginListenerHandle;

addListener(eventName: 'onAdOpened', listenerFunc: (info: any) => void): PluginListenerHandle;

addListener(eventName: 'onAdClosed', listenerFunc: (info: any) => void): PluginListenerHandle;
```

## 📌 INTERSTITIAL

### prepareInterstitial(options: AdOptions): Promise<{ value: boolean }>

```typescript
import { Plugins } from "@capacitor/core";
import { AdOptions } from "capacitor-admob";

const { AdMob } = Plugins;

@Component({
  selector: "admob",
  templateUrl: "admob.component.html",
  styleUrls: ["admob.component.scss"]
})
export class AppComponent {
  options: AdOptions = {
    adId: "Your AD_Id",
    adSize: AdSize.SMART_BANNER,
    position: AdPosition.BOTTOM_CENTER,
    hasTabBar: false, // make it true if you have TabBar Layout.
    tabBarHeight: 56 // you can assign custom margin in pixel default is 56
  };

  constructor() {
    // Prepare interstitial banner
    AdMob.prepareInterstitial(this.options).then(
      value => {
        console.log(value); // true
      },
      error => {
        console.error(error); // show error
      }
    );

    // Subscibe Banner Event Listener
    AdMob.addListener("onInterstitialAdLoaded", (info: boolean) => {
      // You can call showInterstitial() here or anytime you want.

      console.log("Interstitial Ad Loaded");
    });
  }
}
```

### showInterstitial(): Promise<{ value: boolean }>

```typescript
// Show interstitial ad when it’s ready

AdMob.showInterstitial().then(
  value => {
    console.log(value); // true
  },
  error => {
    console.error(error); // show error
  }
);
```

### Event Listener

This following Event Listener can be called in **Interstitial AD**

```typescript
addListener(eventName: 'onInterstitialAdLoaded', listenerFunc: (info: any) => void): PluginListenerHandle;

addListener(eventName: 'onInterstitialAdFailedToLoad', listenerFunc: (info: any) => void): PluginListenerHandle;

addListener(eventName: 'onInterstitialAdOpened', listenerFunc: (info: any) => void): PluginListenerHandle;

addListener(eventName: 'onInterstitialAdLeftApplication', listenerFunc: (info: any) => void): PluginListenerHandle;

addListener(eventName: 'onInterstitialAdClosed', listenerFunc: (info: any) => void): PluginListenerHandle;
```

## 📌 RewardVideo

### prepareRewardVideoAd(options: AdOptions): Promise<{ value: boolean }>

```typescript
import { Plugins } from "@capacitor/core";
import { AdOptions } from "capacitor-admob";

const { AdMob } = Plugins;

@Component({
  selector: "admob",
  templateUrl: "admob.component.html",
  styleUrls: ["admob.component.scss"]
})
export class AAdMobComponent {
  options: AdOptions = {
    adId: "YOUR ADID"
  };

  constructor() {
    // Prepare ReWardVideo
    AdMob.prepareRewardVideoAd(this.options).then(
      value => {
        console.log(value); // true
      },
      error => {
        console.error(error); // show error
      }
    );

    // Subscibe ReWardVideo Event Listener
    AdMob.addListener("onRewardedVideoAdLoaded", (info: boolean) => {
      // You can call showRewardVideoAd() here or anytime you want.
      console.log("RewardedVideoAd Loaded");
    });
  }
}
```

### showRewardVideoAd(): Promise<{ value: boolean }>

```typescript
// Show a RewardVideo AD

AdMob.showRewardVideoAd().then(
  value => {
    console.log(value); // true
  },
  error => {
    console.error(error); // show error
  }
);
```

### pauseRewardedVideo(): Promise<{ value: boolean }>

```typescript
// Pause a RewardVideo AD

AdMob.pauseRewardedVideo().then(
  value => {
    console.log(value); // true
  },
  error => {
    console.error(error); // show error
  }
);
```

### resumeRewardedVideo(): Promise<{ value: boolean }>

```typescript
// Resume a RewardVideo AD

AdMob.resumeRewardedVideo().then(
  value => {
    console.log(value); // true
  },
  error => {
    console.error(error); // show error
  }
);
```

### stopRewardedVideo(): Promise<{ value: boolean }>

```typescript
// Stop a RewardVideo AD

AdMob.stopRewardedVideo().then(
  value => {
    console.log(value); // true
  },
  error => {
    console.error(error); // show error
  }
);
```

### Event Listener

This following Event Listener can be called in **RewardedVideo**

```typescript
addListener(eventName: 'onRewardedVideoAdLoaded', listenerFunc: (info: any) => void): PluginListenerHandle;

addListener(eventName: 'onRewardedVideoAdOpened', listenerFunc: (info: any) => void): PluginListenerHandle;

addListener(eventName: 'onRewardedVideoStarted', listenerFunc: (info: any) => void): PluginListenerHandle;

addListener(eventName: 'onRewardedVideoAdClosed', listenerFunc: (info: any) => void): PluginListenerHandle;

addListener(eventName: 'onRewarded', listenerFunc: (info: any) => void): PluginListenerHandle;

addListener(eventName: 'onRewardedVideoAdLeftApplication', listenerFunc: (info: any) => void): PluginListenerHandle;

addListener(eventName: 'onRewardedVideoAdFailedToLoad', listenerFunc: (info: any) => void): PluginListenerHandle;

addListener(eventName: 'onRewardedVideoCompleted', listenerFunc: (info: any) => void): PluginListenerHandle;
```

# API

### 📌 AdOptions

```typescript
interface AdOptions {
  adId: string;
  adSize?: AdSize;
  position?: AdPosition;
  hasTabBar?: boolean; // optional: default false
  tabBarHeight?: number; // set cutom height in pixal default is 56
  userId?: string; // RewardedVideo ONLY, Optional user ID useful when using SSV// Height in Pixal
}
```

### 📌 AdSize

```typescript
enum AdSize {
  BANNER = "BANNER",

  FLUID = "FLUID",

  FULL_BANNER = "FULL_BANNER",

  LARGE_BANNER = "LARGE_BANNER",

  LEADERBOARD = "LEADERBOARD",

  MEDIUM_RECTANGLE = "MEDIUM_RECTANGLE",

  SMART_BANNER = "SMART_BANNER",

  CUSTOM = "CUSTOM"
}
```

### 📌 AdPosition

```typescript
enum AdPosition {
  TOP_CENTER = "TOP_CENTER",

  CENTER = "CENTER",

  BOTTOM_CENTER = "BOTTOM_CENTER"
}
```

## Contributing

- 🌟 Star this repository
- 📋 Open issue for feature requests

## Roadmap

- [Capacitor Plugins](https://capacitor.ionicframework.com/docs/plugins/)


CHANGELOG.md
# Changelog

### [v1.0.0](https://github.com/rahadur/capacitor-admob/tree/v1.0.0)
- Fixed Androidx CoordinatorLayout **[#35](https://github.com/rahadur/capacitor-admob/issues/35)**
- Fixed Banner and interstitial share the same Event Listeners **[#40](https://github.com/rahadur/capacitor-admob/issues/40)**
- Fixed AdMob error in Mobile after build **[#48](https://github.com/rahadur/capacitor-admob/issues/48)**
- Fixed Ad overlaps tab bar iOS **[#27](https://github.com/rahadur/capacitor-admob/issues/27)**
- Migrate Demo App to Capacitor 2.1.2

### [v0.1.0](https://github.com/rahadur/capacitor-admob/tree/v0.1.0)

- Implement AdMob iOS SDK.

### [v0.0.9](https://github.com/rahadur/capacitor-admob/tree/v0.0.9)

- Fixed Plugin throws error when trying to show reward video **[#2](https://github.com/rahadur/capacitor-admob/issues/2)**
- Fixed AD overlapping tabs **[#4](https://github.com/rahadur/capacitor-admob/issues/4)**

- Fixed Cause: startup failed **[#7](https://github.com/rahadur/capacitor-admob/issues/7)**

android/src/main/java/app/xplatform/capacitor/plugins/AdMob.java
package app.xplatform.capacitor.plugins;

import android.Manifest;
import android.util.Log;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RelativeLayout;

import androidx.annotation.NonNull;
import androidx.coordinatorlayout.widget.CoordinatorLayout;

import com.getcapacitor.JSObject;
import com.getcapacitor.NativePlugin;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdSize;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.InterstitialAd;
import com.google.android.gms.ads.MobileAds;
import com.google.android.gms.ads.rewarded.RewardItem;
import com.google.android.gms.ads.rewarded.RewardedAd;
import com.google.android.gms.ads.rewarded.RewardedAdCallback;
import com.google.android.gms.ads.rewarded.RewardedAdLoadCallback;


@NativePlugin(
    permissions = {
            Manifest.permission.ACCESS_NETWORK_STATE,
            Manifest.permission.INTERNET
    }
)
public class AdMob extends Plugin {

    private PluginCall call;

    private ViewGroup mViewGroup;


    private RelativeLayout mAdViewLayout;


    private AdView mAdView;


    private InterstitialAd mInterstitialAd;


    private RewardedAd mRewardedVideoAd;


    // Initialize AdMob with appId
    @PluginMethod()
    public void initialize(PluginCall call) {
        /* Sample AdMob App ID: ca-app-pub-3940256099942544~3347511713 */
        String appId = call.getString("appId", "ca-app-pub-3940256099942544~3347511713");

        try {
            MobileAds.initialize(this.getContext(), appId);

            mViewGroup = (ViewGroup) ((ViewGroup) getActivity()
                    .findViewById(android.R.id.content)).getChildAt(0);

            call.success();

        } catch (Exception ex) {
            call.error(ex.getLocalizedMessage(), ex);
        }
    }


    // Show a banner Ad
    @PluginMethod()
    public void showBanner(PluginCall call) {
        /* Dedicated test ad unit ID for Android banners: ca-app-pub-3940256099942544/6300978111*/
        String adId = call.getString("adId", "ca-app-pub-3940256099942544/6300978111");
        String adSize = call.getString("adSize", "SMART_BANNER");
        String adPosition = call.getString("position", "BOTTOM_CENTER");

        try {

            if (mAdView == null) {
                mAdView = new AdView(getContext());
                mAdView.setAdUnitId(adId);


                Log.d(getLogTag(), "Ad ID: " + adId);


                switch (adSize) {
                    /*case "SMART_BANNER":
                        mAdView.setAdSize(AdSize.SMART_BANNER);
                        break;*/
                    case "BANNER":
                        mAdView.setAdSize(AdSize.BANNER);
                        break;
                    case "FLUID":
                        mAdView.setAdSize(AdSize.FLUID);
                        break;
                    case "FULL_BANNER":
                        mAdView.setAdSize(AdSize.FULL_BANNER);
                        break;
                    case "LARGE_BANNER":
                        mAdView.setAdSize(AdSize.LARGE_BANNER);
                        break;
                    case "LEADERBOARD":
                        mAdView.setAdSize(AdSize.LEADERBOARD);
                        break;
                    case "MEDIUM_RECTANGLE":
                        mAdView.setAdSize(AdSize.MEDIUM_RECTANGLE);
                        break;
                    default:
                        mAdView.setAdSize(AdSize.SMART_BANNER);
                        break;
                }
            }


            // Setup AdView Layout

            mAdViewLayout = new RelativeLayout(getContext());
            mAdViewLayout.setHorizontalGravity(Gravity.CENTER_HORIZONTAL);
            mAdViewLayout.setVerticalGravity(Gravity.BOTTOM);

            final CoordinatorLayout.LayoutParams mAdViewLayoutParams = new CoordinatorLayout.LayoutParams(
                    CoordinatorLayout.LayoutParams.WRAP_CONTENT,
                    CoordinatorLayout.LayoutParams.WRAP_CONTENT
            );


            switch (adPosition) {
                case "TOP_CENTER":
                    mAdViewLayoutParams.gravity = Gravity.TOP;
                    break;
                case "CENTER":
                    mAdViewLayoutParams.gravity = Gravity.CENTER;
                    break;
                case "BOTTOM_CENTER":
                    mAdViewLayoutParams.gravity = Gravity.BOTTOM | Gravity.CENTER;
                    break;
                default:
                    mAdViewLayoutParams.gravity = Gravity.BOTTOM;
                    break;
            }


            // Set Bottom margin for TabBar
            boolean hasTabBar = call.getBoolean("hasTabBar", false);
            if (hasTabBar) {
                float density = getContext().getResources().getDisplayMetrics().density;
                float tabBarHeight = call.getInt("tabBarHeight", 56);
                int margin = (int) (tabBarHeight * density);
                mAdViewLayoutParams.setMargins(0, 0, 0, margin);
            }


            // Remove child from AdViewLayout
            getActivity().runOnUiThread(new Runnable() {
                @Override
                public void run() {
                if (mAdView.getParent() != null) {
                    ((ViewGroup) mAdView.getParent()).removeView(mAdView);
                }
                mAdViewLayout.setLayoutParams(mAdViewLayoutParams);
                // Add AdView into AdViewLayout
                mAdViewLayout.addView(mAdView);
                }
            });


            // Run AdMob In Main UI Thread
            getActivity().runOnUiThread(new Runnable() {
                @Override
                public void run() {
                mAdView.loadAd(new AdRequest.Builder().build());

                mAdView.setAdListener(new AdListener() {
                    @Override
                    public void onAdLoaded() {
                        notifyListeners("onAdLoaded", new JSObject().put("value", true));
                        super.onAdLoaded();
                    }

                    @Override
                    public void onAdFailedToLoad(int i) {
                        notifyListeners("onAdFailedToLoad", new JSObject().put("errorCode", i));
                        super.onAdFailedToLoad(i);
                    }

                    @Override
                    public void onAdOpened() {
                        notifyListeners("onAdOpened", new JSObject().put("value", true));
                        super.onAdOpened();
                    }

                    @Override
                    public void onAdClosed() {
                        notifyListeners("onAdClosed", new JSObject().put("value", true));
                        super.onAdClosed();
                    }
                });

                // Add AdViewLayout top of the WebView
                mViewGroup.addView(mAdViewLayout);
                }
            });

            call.success(new JSObject().put("value", true));

        } catch (Exception ex) {
            call.error(ex.getLocalizedMessage(), ex);
        }
    }


    // Hide the banner, remove it from screen, but can show it later
    @PluginMethod()
    public void hideBanner(PluginCall call) {
        try {
            getActivity().runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    if (mAdViewLayout != null) {
                        mAdViewLayout.setVisibility(View.GONE);
                        mAdView.pause();
                    }
                }
            });

            call.success(new JSObject().put("value", true));

        } catch (Exception ex) {
            call.error(ex.getLocalizedMessage(), ex);
        }
    }


    // Resume the banner, show it after hide
    @PluginMethod()
    public void resumeBanner(PluginCall call) {
        try {
            getActivity().runOnUiThread(new Runnable() {
                @Override
                public void run() {
                if (mAdViewLayout != null && mAdView != null) {
                    mAdViewLayout.setVisibility(View.VISIBLE);
                    mAdView.resume();
                    Log.d(getLogTag(), "Banner AD Resumed");
                }
                }
            });

            call.success(new JSObject().put("value", true));

        } catch (Exception ex) {
            call.error(ex.getLocalizedMessage(), ex);
        }
    }


    // Destroy the banner, remove it from screen.
    @PluginMethod()
    public void removeBanner(PluginCall call) {
        try {
            if (mAdView != null) {
                getActivity().runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        if (mAdView != null) {
                            mViewGroup.removeView(mAdViewLayout);
                            mAdViewLayout.removeView(mAdView);
                            mAdView.destroy();
                            mAdView = null;
                            Log.d(getLogTag(), "Banner AD Removed");
                        }
                    }
                });
            }

            call.success(new JSObject().put("value", true));

        } catch (Exception ex) {
            call.error(ex.getLocalizedMessage(), ex);
        }
    }


    // Prepare interstitial Ad
    @PluginMethod()
    public void prepareInterstitial(final PluginCall call) {
        this.call = call;
        /* dedicated test ad unit ID for Android interstitials:
            ca-app-pub-3940256099942544/1033173712
        */
        String adId = call.getString("adId", "ca-app-pub-3940256099942544/1033173712");


        try {

            mInterstitialAd = new InterstitialAd(getContext());
            mInterstitialAd.setAdUnitId(adId);


            getActivity().runOnUiThread(new Runnable() {
                @Override
                public void run() {
                mInterstitialAd.loadAd(new AdRequest.Builder().build());

                mInterstitialAd.setAdListener(new AdListener() {
                    @Override
                    public void onAdLoaded() {
                        // Code to be executed when an ad finishes loading.
                        notifyListeners("onInterstitialAdLoaded", new JSObject().put("value", true));
                        call.success(new JSObject().put("value", true));
                        super.onAdLoaded();

                    }

                    @Override
                    public void onAdFailedToLoad(int errorCode) {
                        // Code to be executed when an ad request fails.
                        notifyListeners("onInterstitialAdFailedToLoad", new JSObject().put("errorCode", errorCode));
                        super.onAdFailedToLoad(errorCode);
                    }

                    @Override
                    public void onAdOpened() {
                        // Code to be executed when the ad is displayed.
                        notifyListeners("onInterstitialAdOpened", new JSObject().put("value", true));
                        super.onAdOpened();
                    }

                    @Override
                    public void onAdLeftApplication() {
                        // Code to be executed when the user has left the app.
                        notifyListeners("onInterstitialAdLeftApplication", new JSObject().put("value", true));
                        super.onAdLeftApplication();
                    }

                    @Override
                    public void onAdClosed() {
                        // Code to be executed when when the interstitial ad is closed.
                        notifyListeners("onInterstitialAdClosed", new JSObject().put("value", true));
                        super.onAdClosed();
                    }
                });

                }
            });

        } catch (Exception ex) {
            call.error(ex.getLocalizedMessage(), ex);
        }
    }


    // Show interstitial Ad
    @PluginMethod()
    public void showInterstitial(final PluginCall call) {
        try {
            getActivity().runOnUiThread(new Runnable() {
                @Override
                public void run() {
                if (mInterstitialAd != null && mInterstitialAd.isLoaded()) {
                    getActivity().runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            mInterstitialAd.show();
                        }
                    });
                    call.success(new JSObject().put("value", true));
                } else {
                    call.error("The interstitial wasn't loaded yet.");
                }
                }
            });
        } catch (Exception ex) {
            call.error(ex.getLocalizedMessage(), ex);
        }
    }


    // Prepare a RewardVideoAd
    @PluginMethod()
    public void prepareRewardVideoAd(final PluginCall call) {
        this.call = call;
        /* dedicated test ad unit ID for Android rewarded video:
            ca-app-pub-3940256099942544/5224354917
        */
        final String adId = call.getString("adId", "ca-app-pub-3940256099942544/5224354917");

        try {
            //mRewardedVideoAd = MobileAds.getRewardedVideoAdInstance(getContext());
            mRewardedVideoAd = new RewardedAd(getContext(), adId);

            getActivity().runOnUiThread(new Runnable() {
                @Override
                public void run() {
                //mRewardedVideoAd.loadAd(adId, new AdRequest.Builder().build());

                mRewardedVideoAd.loadAd(new AdRequest.Builder().build(), new RewardedAdLoadCallback() {
                    @Override
                    public void onRewardedAdLoaded() {
                        // Ad successfully loaded.
                        call.success(new JSObject().put("value", true));
                        notifyListeners("onRewardedVideoAdLoaded", new JSObject().put("value", true));
                    }

                    @Override
                    public void onRewardedAdFailedToLoad(int errorCode) {
                        // Ad failed to load.
                        notifyListeners("onRewardedVideoAdFailedToLoad", new JSObject().put("value", true));
                    }
                });
                }
            });


        } catch (Exception ex) {
            call.error(ex.getLocalizedMessage(), ex);
        }

    }

    // Show a RewardVideoAd
    @PluginMethod()
    public void showRewardVideoAd(final PluginCall call) {
        try {
            getActivity().runOnUiThread(new Runnable() {
                @Override
                public void run() {
                if (mRewardedVideoAd != null && mRewardedVideoAd.isLoaded()) {
                    getActivity().runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                        // mRewardedVideoAd.show();
                        mRewardedVideoAd.show(getActivity(), new RewardedAdCallback() {
                            @Override
                            public void onRewardedAdOpened() {
                                // Ad opened.
                                notifyListeners("onRewardedVideoAdOpened", new JSObject().put("value", true));
                            }

                            @Override
                            public void onRewardedAdClosed() {
                                // Ad closed.
                                notifyListeners("onRewardedVideoAdClosed", new JSObject().put("value", true));
                            }

                            @Override
                            public void onUserEarnedReward(@NonNull RewardItem reward) {
                                // User earned reward.
                                notifyListeners("onRewarded", new JSObject().put("value", reward));
                            }

                            @Override
                            public void onRewardedAdFailedToShow(int errorCode) {
                                // Ad failed to display.
                                notifyListeners("onRewardedVideoAdFailedToShow", new JSObject().put("value", errorCode));
                            }
                        });
                        }
                    });
                    call.success(new JSObject().put("value", true));
                } else {
                    call.error("The RewardedVideoAd wasn't loaded yet.");
                }
                }
            });

        } catch (Exception ex) {
            call.error(ex.getLocalizedMessage(), ex);
        }
    }


    // Pause a RewardVideoAd
    @Deprecated
    @PluginMethod()
    public void pauseRewardedVideo(PluginCall call) {
        /*try {
            getActivity().runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    mRewardedVideoAd.pause(getContext());
                }
            });
            call.success(new JSObject().put("value", true));
        } catch (Exception ex) {
            call.error(ex.getLocalizedMessage(), ex);
        }*/
        call.error("This method no longer support by AdMob SDK");
    }


    // Resume a RewardVideoAd
    @Deprecated
    @PluginMethod()
    public void resumeRewardedVideo(PluginCall call) {
        /*try {
            getActivity().runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    mRewardedVideoAd.resume(getContext());
                }
            });
            call.success(new JSObject().put("value", true));
        } catch (Exception ex) {
            call.error(ex.getLocalizedMessage(), ex);
        }*/
        call.error("This method no longer support by AdMob SDK");
    }


    // Destroy a RewardVideoAd
    @Deprecated
    @PluginMethod()
    public void stopRewardedVideo(PluginCall call) {
        /*try {
            getActivity().runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    mRewardedVideoAd.destroy(getContext());
                }
            });
            call.success(new JSObject().put("value", true));
        } catch (Exception ex) {
            call.error(ex.getLocalizedMessage(), ex);
        }*/
        call.error("This method no longer support by AdMob SDK");
    }
}
